cpp/src/dictionary/replace.cu

```diff
@@ -0,0 +1,242 @@
+  }();
+
+  // take the keys from the matched column (allocated above using mr parameter)
+  std::unique_ptr<column> keys_column(std::move(input_matched->release().children.back()));
+
+  // create column with keys_column and indices_column
+  return make_dictionary_column(std::move(keys_column),
+                                std::move(indices_column),
+                                std::move(*(contents.null_mask.release())),
+                                null_count);
+}
+
+/**
+ * @copydoc cudf::dictionary::detail::replace_nulls(cudf::column_view const&,cudf::scalar
+ * const&,rmm::mr::device_memory_resource*,cudaStream_t)
+ */
+std::unique_ptr<column> replace_nulls(dictionary_column_view const& input,
+                                      scalar const& replacement,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+{
+  if (input.size() == 0) { return cudf::empty_like(input.parent()); }
+  if (!input.has_nulls() || !replacement.is_valid()) {
+    return std::make_unique<cudf::column>(input.parent());
+  }
+  CUDF_EXPECTS(input.keys().type() == replacement.type(), "keys must match scalar type");
+
+  // first add the replacment to the keys so only the indices need to be replaced
+  auto input_matched = dictionary::detail::add_keys(
+    input,
+    make_column_from_scalar(replacement, 1, rmm::mr::get_current_device_resource(), stream)->view(),
+    mr,
+    stream);
+  auto const input_view = dictionary_column_view(input_matched->view());
+  auto scalar_index =
+    get_index(input_view, replacement, rmm::mr::get_current_device_resource(), stream);
+
+  // now build the new indices by doing replace-null on the updated indices
+  auto const input_indices = input_view.get_indices_annotated();
+  auto new_indices =
+    replace_indices(input_indices, make_scalar_iterator(*scalar_index), mr, stream);
+
+  // create the indices for the output column
+  auto const new_type     = get_indices_type_for_size(input_view.keys().size());
+  auto const indices_type = new_indices->type();
+  auto contents           = new_indices->release();
+  // build the indices for the output column
+  auto indices_column = [&] {
+    if (new_type.id() == cudf::type_id::UINT32) {
+      return std::make_unique<column>(cudf::data_type{cudf::type_id::UINT32},
+                                      input_view.size(),
+                                      std::move(*(contents.data.release())),
+                                      rmm::device_buffer{0, stream, mr},
+                                      0);
+    }
+    cudf::column_view cast_view{indices_type, input_view.size(), contents.data->data()};
+    return cudf::detail::cast(cast_view, new_type, mr, stream);
+  }();
+
+  // take the keys from the matched column (allocated above using mr parameter)
+  std::unique_ptr<column> keys_column(std::move(input_matched->release().children.back()));
```

Comment:
Seems to be a lot of duplication with the non-scalar version. Perhaps factor it out?